_posts/2015-01-13-matryoshka-configurable-caching-library-for-php.md

```diff
@@ -0,0 +1,147 @@
+---
+layout: with-comments
+title: "Matryoshka: A Configurable Caching Library for PHP"
+author: Marc Zych
+author_url: https://github.com/marczych
+summary: We recently open-sourced Matryoshka, a configurable caching library
+         for PHP which makes common operations easier and allows for on-the-fly
+         configuration.
+---
+
+Like most websites, we make heavy use of caching to reduce load on our servers and decrease page response times.
+Our caching daemon of choice is [memcached].
+The PHP extensions are certainly usable and provide all of the core functionality that you could need.
+However, we use a lot of patterns to make our day to day caching much easier that aren't provided by the extensions.
+
+In comes [Matryoshka], an open source caching library for PHP which makes common operations easier and allows for on-the-fly configuration.
+
+# Configurable Behavior
+
+Matryoshka is designed to be very configurable.
+You can add functionality on the fly simply by wrapping an existing `Backend` with a new one.
+We use this extensively to prefix keys, modify expiration times, disable gets, gather metrics, etc.
+
+Start off with a `Memcache` instance:
+
+{% highlight php startinline=true %}
+// From the native extension.
+$memcache = new Memcache();
+$memcache->pconnect('localhost', 11211);
+$cache = Matryoshka\Memcache::create($memcache);
+{% endhighlight %}
+
+Then prefix all the keys:
+
+{% highlight php startinline=true %}
+$prefixedCache = new Matryoshka\Prefix($cache, 'prefix-');
+// The key ends up being "prefix-key".
+$prefixedCache->set('key', 'value');
+$value = $prefixedCache->get('key');
+{% endhighlight %}
+
+Finally double all expiration times for the prefixed backend:
+
+{% highlight php startinline=true %}
+$doubleExpiration = function($expiration) {
+   return $expiration * 2;
+};
+$cache = new Matryoshka\ExpirationChange($prefixedCache,
+ $doubleExpiration);
+// Results in an expiration time of 20 for "prefix-key".
+$cache->set('key', 'value', 10);
+{% endhighlight %}
+
+By using composition, caching configurations can be assembled on the fly quite easily.
+The core API is identical for all backends so the caller doesn't need to be aware of the exact configuration.
+Common configurations and base backends (like memcached connections) can be made into singletons or provided using dependency injection in your application.
+Additionally, this architecture results in very maintainable and testable code because each class has exactly one job.
+
+# Scopes
+
+Cache invalidation is hard.
+To make it easier, Matryoshka provides "cache scopes" to invalidate a group of keys at once.
+[This works][Scope.php] by prefixing all keys with a unique value that is stored in the backend using the scope name.
+
+{% highlight php startinline=true %}
+$cache = new Matryoshka\Scope($memcachedBackend, 'name');
+
+// This results in a get request to memcached for 'scope-name'
+// which results in something like '0fb4ae36'. This `set` call
```

Comment:
What I have now makes sense to me but it can probably be improved. I want to make it clear that the random prefix is stored in memcached which makes it tricky.